### ISSUE TITLE
Broken smerge-mode

### ISSUE THREAD

**Issue by Gravemind** [elp-revive/auto-highlight-symbol#18]
I had a somewhat broken smerge-mode (git diff3 resolution), and, it took some time, but I finally narrowed it down to auto-highlight-symbol commit bc1564a8d5 (thanks git-bisect). I have no idea how it's related to smerge, but it's quite non-mistakable: on `bc1564a8d5` smerge is borken, on `bc1564a8d5~` it is not.

To reproduce the bug, I simply create a `test.py` file (does not happen with a .txt file...):

```py
<<<<<<< A
a
||||||| C
=======
b
>>>>>>> B
```

Enable ahs with `M-x auto-highlight-symbol-mode` (if not already), then `M-x smerge-mode`, then `M-x smerge-keep-upper` and you get:

```py
<<<<<<< A
a
```

instead of:

```py
a
```

All the `smerge-keep-*` seem borken.

If I checkout `auto-highlight-symbol` to `bc1564a8d5~`, then everything works as expected again.

I don't see any error in `*Messages*`, I use emacs 27.1.

EDIT: add `M-x auto-highlight-symbol-mode` repro step

**Comment by Gravemind** [comment]
Of course I just tested with `emacs -Q` and I cannot reproduce it; There must be another plugin involved... I'll keep looking.

**Comment by Gravemind** [comment]
Ok, I did reproduce it with `emacs -Q` (you just need not to forget to run `M-x auto-highlight-symbol-mode` before...)

**Comment by jcowgill** [comment]
I've also noticed `replace-string` is broken after bc1564a. After replacing the first occurence I get the message:

    replace-match-maybe-edit: Match data clobbered by buffer modification hooks

[Stack overflow suggests](https://emacs.stackexchange.com/questions/38800/match-data-clobbered-by-buffer-modification-hooks/38862#38862) this is caused by misuse of the `before-change-functions` and `after-change-functions` hooks.

**Comment by zabe40** [comment]
I noticed this problem as well while trying to `query-replace` some code. I was able to fix it in my config by wrapping `ahs-start-timer` with `save-match-data` via advice like so:
```
(define-advice ahs-start-timer (:around (old-fun &rest rest) save-match-data)
    (save-match-data
      (apply old-fun rest)))
```
The specific offender seems to be `ahs-highlight-p` with it's calls to `bounds-of-thing-at-point`, and `ahs-symbol-p`. It seems to me like the proper fix is just to wrap the body of `ahs-start-timer` with `save-match-data`.

The Emacs manual does mention this on the page about buffer change functions:

> The functions you use in these hooks should save and restore the match data if they do anything that uses regular expressions; otherwise, they will interfere in bizarre ways with the editing operations that call them.

but IMO this should be mentioned directly in the docstrings themselves.
